src/overlays/actors/ovl_Elf_Msg3/z_elf_msg3.c

```diff
@@ -29,28 +29,122 @@ const ActorInit Elf_Msg3_InitVars = {
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg3/ElfMsg3_SetupAction.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg3/func_80A2CD1C.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg3/ElfMsg3_Init.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg3/ElfMsg3_Destroy.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg3/func_80A2CF50.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg3/func_80A2CF7C.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg3/ElfMsg3_Update.s")
+void ElfMsg3_SetupAction(ElfMsg3* this, ElfMsg3ActionFunc actionFunc) {
+    (void)"共倒れ"; // "Collapse together"
+    (void)"共倒れ";
+    this->actionFunc = actionFunc;
+}
+
+s32 func_80A2CD1C(ElfMsg3* this, GlobalContext* globalCtx) {
+    if ((this->actor.home.rot.y > 0) && (this->actor.home.rot.y < 0x81) &&
+        (Flags_GetSwitch(globalCtx, this->actor.home.rot.y - 1) != 0)) {
+        if (ELFMSG3_GET_SWITCH(&this->actor) != 0x7F) {
+            Flags_SetSwitch(globalCtx, ELFMSG3_GET_SWITCH(&this->actor));
+        }
+        Actor_MarkForDeath(&this->actor);
+        return true;
+    }
+    if (this->actor.home.rot.y == 0x81) {
+        if (Flags_GetClear(globalCtx, this->actor.room) != 0) {
+            if (ELFMSG3_GET_SWITCH(&this->actor) != 0x7F) {
+                Flags_SetSwitch(globalCtx, ELFMSG3_GET_SWITCH(&this->actor));
+            }
+            Actor_MarkForDeath(&this->actor);
+            return true;
+        }
+    }
+    if (ELFMSG3_GET_SWITCH(&this->actor) == 0x7F) {
+        return false;
+    }
+    if (Flags_GetSwitch(globalCtx, ELFMSG3_GET_SWITCH(&this->actor)) != 0) {
+        Actor_MarkForDeath(&this->actor);
+        return true;
+    }
+    return false;
+}
+
+void ElfMsg3_Init(Actor* thisx, GlobalContext* globalCtx) {
+    ElfMsg3* this = THIS;
+    s32 absolute;
+
+    if (func_80A2CD1C(this, globalCtx) == 0) {
+        Actor_ProcessInitChain(&this->actor, sInitChain);
+        if (ABS_ALT(this->actor.home.rot.x) == 0) {
+            this->actor.scale.z = 0.4f;
+            this->actor.scale.x = 0.4f;
+        } else {
+            absolute = ABS_ALT(this->actor.home.rot.x);
+            this->actor.scale.z = absolute * 0.04f;
+            this->actor.scale.x = absolute * 0.04f;
+        }
+        if (this->actor.home.rot.z == 0) {
+            this->actor.scale.y = 0.4f;
+        } else {
+            this->actor.scale.y = this->actor.home.rot.z * 0.04f;
+        }
+        ElfMsg3_SetupAction(this, func_80A2CF7C);
+        this->actor.shape.rot.z = 0;
+        this->actor.shape.rot.y = this->actor.shape.rot.z;
+        this->actor.shape.rot.x = this->actor.shape.rot.z;
+    }
+}
+
+void ElfMsg3_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+}
+
+s32 func_80A2CF50(ElfMsg3* this) {
+    if (ELFMSG3_GET_8000(&this->actor) != 0) {
+        return (ELFMSG3_GET_FF(&this->actor)) + 0x200;
+    }
+    return -0x200 - (ELFMSG3_GET_FF(&this->actor));
```

Comment:
IMO this would look better with an else, preferably with the 0x200s in the same position (one way may match and the other not).

Reply:
can you give me an example?

Reply:
```suggestion
    if (ELFMSG3_GET_8000(&this->actor) != 0) {
        return 0x200 + ELFMSG3_GET_FF(&this->actor);
    } else
        return -0x200 - ELFMSG3_GET_FF(&this->actor);
    }
```
or with the addends reversed if it matches.